<?xml version="1.0"?>
<robot  name="gripper_robot">

<gazebo reference="base_link">
    <material>Gazebo/Grey</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>
  <!-- Link1 -->
  <gazebo reference="link_1">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <!-- Link2 -->
  <gazebo reference="link_2">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <!-- Link3 -->
  <gazebo reference="link_3">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <!-- Link4 -->
  <gazebo reference="link_4">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>

  <gazebo reference="link5">
    <material>Gazebo/Grey</material>
    <kp>1000000.0</kp>
    <kd>1.0</kd>
    <mu1>1000.0</mu1>
    <mu2>1000.0</mu2>
  </gazebo>
  <gazebo reference="link6">
    <material>Gazebo/Grey</material>
    <kp>1000000.0</kp>
    <kd>1.0</kd>
    <mu1>1000.0</mu1>
    <mu2>1000.0</mu2>
  </gazebo>

<!-- only movable links have number in the name, like link1, link2 -->
<!-- only movable links have number in the name, like joint1, joint2 -->

<!-- world link -->
<link name="world"/>

<!-- base_link and its fixed joint -->
<joint name="joint_fix" type="fixed">
	<parent link="world"/>
	<child link="base_link"/>
	<dynamics damping="10" friction="1.0"/>
</joint>

<link name="base_link">
	<collision>
		<origin xyz="0 0 0.5" rpy="0 0 0"/>
		<geometry>
			<box size="0.2 0.2 0.7"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.5" rpy="0 0 0"/>
		<geometry>
			<box size="0.2 0.2 1"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.5" rpy="0 0 0"/>
		<mass value="1"/>
		<inertia
			ixx="1" ixy="0.0" ixz="0.0"
			iyy="1" iyz="0.0"
			izz="1"/>
	</inertial>
</link>

<!-- link1 and its revolute joint1 relative to base_link -->
<!-- rotating along z axis -->
<joint name="joint1" type="revolute">
	<parent link="base_link"/>
	<child link="link1"/>
	<origin xyz="0 0 1" rpy="0 0 0"/>
	<axis xyz="0 0 1"/>
	<limit effort="100" velocity="0.5" lower="-1.5708" upper="1.5708"/>
	<dynamics damping="10" friction="1.0"/>
</joint>

<link name="link1">
	<collision>
		<origin xyz="0 0 0.05" rpy="0 0 0"/>
		<geometry>
			<cylinder length="0.1" radius="0.1"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.05" rpy="0 0 0"/>
		<geometry>
			<cylinder length="0.1" radius="0.1"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.05" rpy="0 0 0"/>
		<mass value="1"/>
		<inertia
			ixx="0.001" ixy="0.0" ixz="0.0"
			iyy="0.001" iyz="0.0"
			izz="0.001"/>
	</inertial>
</link>

<gazebo reference="link1">
	<material>Gazebo/Grey</material>
</gazebo>

<!-- link2 and its revolute joint2 relative to link1 -->
<!-- rotating along y axis -->
<joint name="joint2" type="revolute">
	<parent link="link1"/>
	<child link="link2"/>
	<origin xyz="0 0 0.1" rpy="0 0 0"/>
	<axis xyz="0 1 0"/>
	<limit effort="100" velocity="0.5" lower="-0.1" upper="1.5708"/>
	<dynamics damping="10" friction="1.0"/>
</joint>

<link name="link2">
	<collision>
		<origin xyz="0 0 0.5" rpy="0 0 0"/>
		<geometry>
			<cylinder length="1" radius="0.1"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.5" rpy="0 0 0"/>
		<geometry>
			<cylinder length="1" radius="0.1"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.5" rpy="0 0 0"/>
		<mass value="1"/>
		<inertia
			ixx="0.1" ixy="0.0" ixz="0.0"
			iyy="0.1" iyz="0.0"
			izz="0.005"/>
	</inertial>
</link>

<gazebo reference="link2">
	<material>Gazebo/Grey</material>
</gazebo>

<!-- link3 and its revolute joint3 relative to link2 -->
<!-- rotating along y axis -->
<joint name="joint3" type="revolute">
	<parent link="link2"/>
	<child link="link3"/>
	<origin xyz="0 0 1" rpy="0 0 0"/>
	<axis xyz="0 1 0"/>
	<limit effort="10" velocity="0.5" lower="-0.1" upper="3.1416"/>
	<dynamics damping="10" friction="1.0"/>
</joint>

<link name="link3">
	<collision>
		<origin xyz="0 0 0.5" rpy="0 0 0"/>
		<geometry>
			<cylinder length="1" radius="0.1"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.5" rpy="0 0 0"/>
		<geometry>
			<cylinder length="1" radius="0.1"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.5" rpy="0 0 0"/>
		<mass value="1"/>
		<inertia
			ixx="0.1" ixy="0.0" ixz="0.0"
			iyy="0.1" iyz="0.0"
			izz="0.005"/>
	</inertial>
</link>

<gazebo reference="link3">
	<material>Gazebo/Grey</material>
</gazebo>

<!-- link4 and its revolute joint4 relative to link3 -->
<!-- rotating along y axis -->
<joint name="joint4" type="fixed">
	<parent link="link3"/>
	<child link="link4"/>
	<origin xyz="0 0 1" rpy="0 -1.0 0"/>
	<axis xyz="0 1 0"/>
	<dynamics damping="10" friction="1.0"/>
</joint>

<link name="link4">
	<collision>
		<origin xyz="0 0 0" rpy="1.5708 0 0"/>
		<geometry>
			<cylinder length="0.4" radius="0.04"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0" rpy="1.5708 0 0"/>
		<geometry>
			<cylinder length="0.4" radius="0.04"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0" rpy="1.5708 0 0"/>
		<mass value="0.4"/>
		<inertia
			ixx="0.05" ixy="0.0" ixz="0.0"
			iyy="0.002" iyz="0.0"
			izz="0.05"/>
	</inertial>
</link>



<gazebo reference="link4">
	<material>Gazebo/Red</material>
</gazebo>

<joint name="camera_joint" type="fixed">
    <origin xyz="0.15 0 -0.1" rpy="0 -1.5708 0"/>
    <parent link="link4"/>
    <child link="camera_link"/>
  </joint>

  <!-- Camera -->
  <link name="camera_link">
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.05 0.05 0.05"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    <box size="0.05 0.05 0.05"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>
  
  
<gazebo reference="camera_link">
	<material>Gazebo/Black</material>
</gazebo>

<!-- link5 and its prismatic joint5 relative to link4 -->
<!-- translating along y axis -->
<joint name="joint5" type="prismatic">
	<parent link="link4"/>
	<child link="link5"/>
	<origin xyz="0 0.2 0" rpy="0 0 0"/>
	<axis xyz="0 1 0"/>
	<limit effort="50" velocity="100" lower="-0.16" upper="0.0"/>
	<dynamics damping="10" friction="1.0"/>
</joint>

<link name="link5">
	<collision>
		<origin xyz="0 0 0.32" rpy="0 0 0"/>
		<geometry>
			<box size="0.2 0.05 0.7"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.32" rpy="0 0 0"/>
		<geometry>
			<box size="0.2 0.05 0.7"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.32" rpy="0 0 0"/>
		<mass value="0.2"/>
		<inertia
			ixx="0.02" ixy="0.0" ixz="0.0"
			iyy="0.02" iyz="0.0"
			izz="0.001"/>
	</inertial>
</link>

<gazebo reference="link5">
	<material>Gazebo/Red</material>
</gazebo>

<!-- link6 and its prismatic joint6 relative to link4 -->
<!-- translating along y axis -->
<joint name="joint6" type="prismatic">
	<parent link="link4"/>
	<child link="link6"/>
	<origin xyz="0 -0.2 0" rpy="0 0 0"/>
	<axis xyz="0 1 0"/>
	<limit effort="50" velocity="100" lower="0.0" upper="0.16"/>
	<dynamics damping="10" friction="1.0"/>
</joint>

<link name="link6">
	<collision>
		<origin xyz="0 0 0.32" rpy="0 0 0"/>
		<geometry>
			<box size="0.2 0.0.5 0.7"/>
		</geometry>
	</collision>

	<visual>
		<origin xyz="0 0 0.32" rpy="0 0 0"/>
		<geometry>
			<box size="0.2 0.05 0.7"/>
		</geometry>
	</visual>

	<inertial>
		<origin xyz="0 0 0.32" rpy="0 0 0"/>
		<mass value="0.2"/>
		<inertia
			ixx="0.02" ixy="0.0" ixz="0.0"
			iyy="0.02" iyz="0.0"
			izz="0.001"/>
	</inertial>
</link>

<gazebo reference="link6">
	<material>Gazebo/Red</material>
</gazebo>




<gazebo reference="camera_link">
    <sensor type="camera" name="camera1">
      <update_rate>30.0</update_rate>
      <camera name="head">
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>600</width>
          <height>400</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.02</near>
          <far>300</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <!-- Noise is sampled independently per pixel on each frame.
               That pixel's noise value is added to each of its color
               channels, which at that point lie in the range [0,1]. -->
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <alwaysOn>true</alwaysOn>
        <updateRate>0.0</updateRate>
        <cameraName>rrbot/camera1</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>camera_link</frameName>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
      </plugin>
    </sensor>
  </gazebo>

<ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>

    <joint name="joint1">
      <command_interface name="position">
        <param name="min">-10</param>
        <param name="max">10</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>

    <joint name="joint2">
      <command_interface name="position">
        <param name="min">-10</param>
        <param name="max">10</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="joint3">
      <command_interface name="position">
        <param name="min">-10</param>
        <param name="max">10</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="joint4">
      <command_interface name="position">
        <param name="min">-10</param>
        <param name="max">10</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="joint5">
      <command_interface name="position">
        <param name="min">-10</param>
        <param name="max">10</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    <joint name="joint6">
      <command_interface name="position">
        <param name="min">-10</param>
        <param name="max">10</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
      <state_interface name="effort"/>
    </joint>
    


  </ros2_control>
  
    <gazebo>
      <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
        <robot_sim_type>gazebo_ros2_control/GazeboSystem</robot_sim_type>
        <parameters>/home/aderajew/ros2_ws/src/my_robot/config/joint_trajectory_p_controller.yaml</parameters>
      </plugin>
    </gazebo>
    


</robot>

